src/algorithm/TraversalIterator.java

```diff
@@ -0,0 +1,115 @@
+package algorithm;
+
+import graph.Graph;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import java.util.Stack;
+
+public class TraversalIterator implements Iterator<Integer> {
+
+  private final Graph graph;
+  private final int start;
+  private Collection<Integer> unvisited;
+  private Collection<Integer> queue;
+  private boolean isBFS;
+  private int count;
+
+  public TraversalIterator(Graph graph, int start, String method) {
+
+    method = method.toLowerCase();
+    boolean isBFS;
+    if (method.equals("bfs")) {
+      isBFS = true;
+    } else if (method.equals("dfs")) {
+      isBFS = false;
+    } else {
+      throw new IllegalArgumentException("Unknown method: " + method);
+    }
+
+    if (!graph.contains(start)) {
+      throw new NoSuchElementException();
+    }
+
+    this.graph = graph;
+    this.start = start;
+    this.unvisited = graph.getVertexIDs();
+    this.isBFS = isBFS;
+    this.count = 0;
+
+    if (isBFS) {
+      this.queue = new LinkedList<>();
+      add(start);
+    } else {
+      this.queue = new Stack<>();
+      push(start);
+    }
+
+    this.unvisited.remove(start);
+  }
+
+  public Graph getGraph() {
+    return graph;
+  }
+
+  public int getStart() {
+    return start;
+  }
+
+  @Override
+  public Integer next() {
+    if (queue.isEmpty()) {
+      int newStart = this.unvisited.iterator().next();
```

Comment:
Take care, if the unvisited is empty(), this would throw Exception. Better verify if this.unvisited.isEmpty(). I know normally you would call hasNext before next, but since this is a public API, you can not stop the user from calling next in any way.